### PR TITLE
Add methods and associations to models

### PR DESCRIPTION
**Pull request by Daniel-Kigozi**
In the following milestone, I did the following
-Add associations to models
-Add methods to users, posts, comments and likes models

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -2,4 +2,8 @@ class User < ApplicationRecord
   has_many :posts, foreign_key: 'author_id'
   has_many :comments
   has_many :likes
```
Comment (Qoosim) [inline app/models/user.rb]: - [ ] Great job adding association to your models. the comments and the likes also required foreign-keys like you did for the posts. It should look something similar to this.

```
  has_many :posts, foreign_key: 'author_id'
  has_many :comments, foreign_key: 'author_id'
  has_many :likes, foreign_key: 'author_id'
```

File: app/models/post.rb [hunk 2]
```diff
@@ -1,5 +1,20 @@
...
+  after_create :increment_user_posts_counter
+  after_destroy :decrement_user_posts_counter
+
+  def increment_user_posts_counter
+    author.increment!(:posts_counter)
+  end
+
+  def decrement_user_posts_counter
+    author.decrement!(:posts_counter)
+  end
```
Comment (Qoosim) [inline app/models/post.rb]: - [OPTIONAL] Great work for using one of the [Rails Active Callbacks](https://guides.rubyonrails.org/active_record_callbacks.html). There's a better way with lesser line of code.
- You can use `after_save` to call the `#update_post_counter` method every time a new post has been saved to database, then make the `update_post_counter` method private.
-  It should look something similar to this. :point_down: 

```
 # app/models/post.rb
  after_save :update_post_counter

  private

  def update_post_counter
    author.increment!(:posts_counter)
  end
```

File: app/models/comment.rb [hunk 3]
```diff
@@ -1,4 +1,15 @@
 class Comment < ApplicationRecord
   belongs_to :user
   belongs_to :post
+
+  after_create :increment_post_comments_counter
+  after_destroy :decrement_post_comments_counter
+
+  def increment_post_comments_counter
+    post.increment!(:comments_counter)
+  end
```
Comment (Qoosim) [inline app/models/comment.rb]: [OPTIONAL] Great work for using one of the [Rails Active Callbacks](https://guides.rubyonrails.org/active_record_callbacks.html). There's a better way with lesser line of code.
- You can use `after_save` to call the `#update_comment_counter` method every time a new comment has been saved to database, then make the `update_comment_counter` method private.
-  It should look something similar to this. :point_down: 

```
 # app/models/post.rb
  after_save :update_comment_counter

  private

  def update_comment_counter
    post.increment!(:comments_counter)
  end
```

File: app/models/like.rb [hunk 4]
```diff
@@ -1,4 +1,15 @@
...
+  after_create :increment_post_likes_counter
+  after_destroy :decrement_post_likes_counter
+
+  def increment_post_likes_counter
+    post.increment!(:likes_counter)
+  end
+
+  def decrement_post_likes_counter
+    post.decrement!(:likes_counter)
+  end
```
Comment (Qoosim) [inline app/models/like.rb]: [OPTIONAL] Great work for using one of the [Rails Active Callbacks](https://guides.rubyonrails.org/active_record_callbacks.html). There's a better way with lesser line of code.
- You can use `after_save` to call the `#update_like_counter` method every time a new like has been saved to database, then make the `update_like_counter` method private.
-  It should look something similar to this. :point_down: 

```
 # app/models/like.rb
  after_save :update_like_counter

  private

  def update_like_counter
    post.increment!(:likes_counter)
  end
```

**Review by Qoosim — CHANGES_REQUESTED** [review]
#  *Status: Required Changes* :recycle:

Hi @Daniel-Kigozi, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## Highlights
-  Created association for models. :heavy_check_mark: 
-  Good README file. :heavy_check_mark: 
-  Correct Gitflow. :heavy_check_mark: 




## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by bobb-Rob — APPROVED** [review]
## STATUS: APPROVED ✔️ 🍾 🍾 🍾 🥳 🥳 🥳 

Hi @Daniel-Kigozi, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 



## Highlights
✅ No Linter error
✅ Correct use of Gitflow
✅ Very descriptive Pull Request(PR)

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._a

**Pull request merged**